core/io/xml_parser.cpp

```diff
@@ -205,17 +205,20 @@ void XMLParser::_parse_comment() {
 
 	char *pCommentBegin = P;
 
-	int count = 1;
+	ERR_FAIL_COND_MSG(*P != '-', "Bad XML comment");
+	++P;
+	ERR_FAIL_COND_MSG(*P != '-', "Bad XML comment");
+	++P;
 
 	// move until end of comment reached
-	while (count) {
-		if (*P == '>') {
-			--count;
-		} else if (*P == '<') {
-			++count;
-		}
-
+	for (;;) {
+		while (*P != '-')
+			++P;
+		if (*P != '-')
+			continue;
 		++P;
+		if (*P == '>')
+			break;
```

Comment:
This would allow closing a comment with `--->`, which is not well-formed according to the spec:

> Note that the grammar does not allow a comment ending in `--->`. 

https://www.w3.org/TR/2008/REC-xml-20081126/#sec-comments

Reply:
Very good remark, I will update my patch to include your remark.
 * detect EOF
 * correct to use the correct syntax (to ignore end comment like ---> )
   Comment	   ::=   	'<!--' ((Char - '-') | ('-' (Char - '-')))* '-->'
* For break on nested comments, is not valid XML
   what do you want? 
   (keep this functionality, adding option to disable/enable this functionality - and by default enable nested comments)